### PR TITLE
Fourier nested optimizer

### PR DESCRIPTION
**Pull request by laurgao**
Fourier nested optimizer does the same functionality as the old Fourier implementation when used with `LayerwiseAnsatzOptimizer`, but with added control over hyperparameters `q` and `R` mentioned in the original paper. When used with `q = infinity` and `R > 0`, accuracy can be significantly improved from old fourier. (For 10-12 qubits, 4 layers, and 
R = 5, approx ratio went from ~65% to ~90%.)

Uh oh it looks like there are many files changed, but really, the key ones are `src/python/zquantum/qaoa/parameter_initialization/_fourier.py` and `tests/zquantum/qaoa/parameter_initialization/fourier_test.py`. Everything else comes from reorganizing code and renaming files from the other interp/fourier PR.

### CODE REVIEW

File: src/python/zquantum/qaoa/parameter_initialization/_fourier.py [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+            inner_optimizer: optimizer used for optimization of parameters
+                after adding a new layer to the ansatz.
+            min_layer: starting number of layers.
+            max_layer: maximum number of layers, at which optimization should stop.
+            n_layers_per_iteration: number of layers added for each iteration.
+            q: length of each of the u and v parameters. Can be any positive integer or
+                infinity. If q = infinity, then q = n_layers and grows unbounded. The
+                authors of the original paper used q = infinity.
+            R: the number of random perturbations we add to the parameters so that we
+                can sometimes escape a local optimum. Can be any integer 0 or above. The
```
Comment (mstechly) [inline src/python/zquantum/qaoa/parameter_initialization/_fourier.py:65]: Super minor style suggestion
```suggestion
                can sometimes escape a local optimum. Can be any non-negative integer. The
```

File: src/python/zquantum/qaoa/parameter_initialization/_fourier.py [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+        return u_v_cost_function
+
+    def _get_u_v_for_next_layer(self, u_v: np.ndarray) -> np.ndarray:
+        """When q = infinity, u_v is extended at the increment of each layer such that
+        the length of u and v is equal to length of gamma and beta. See equation B3
+        of the original paper.
+
+        """
+        # Increment the length of u and v if q = infinity
+        if self._q == np.inf:
```
Comment (mstechly) [inline src/python/zquantum/qaoa/parameter_initialization/_fourier.py:239]: I think we should change it from inf to `None`, that would be more pythonic.

File: src/python/zquantum/qaoa/parameter_initialization/_fourier.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+
+def _perturb_params_randomly(u_v: np.ndarray, alpha: float = 0.6) -> np.ndarray:
+    """Performs one random perturbation.
+
+    The perturbations are sampled from normal distributions with mean of 0 and variance
+    given by `u` and `v`. See Equation B5 or last paragraph of pg. 17
+
+    Alpha is a free parameter corresponding to the strength of the perturbation. A value
+    of 0.6 is what was found to work best by the authors of the original paper and is
+    what we use in this implementation of Fourier.
```
Comment (mstechly) [inline src/python/zquantum/qaoa/parameter_initialization/_fourier.py:309]:
```suggestion
    what we use as default in this implementation of Fourier.
```

**Review by mstechly — CHANGES_REQUESTED** [review]

File: tests/zquantum/qaoa/parameter_initialization/fourier_test.py [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+        if hasattr(cost_function, "gradient"):
+            result.gradient_history = cost_function.gradient(initial_params)
+
+        return result
+
+    inner_optimizer._minimize = custom_minimize
+    return inner_optimizer
+
+
+class TestFouier:
```
Comment (dexter2206) [inline tests/zquantum/qaoa/parameter_initialization/fourier_test.py:82]: Typo :) `TestFouier -> TestFourier`

**Review by dexter2206 — CHANGES_REQUESTED** [review]

File: src/python/zquantum/qaoa/estimators/_cvar_estimator.py [hunk 5]
```diff
@@ -213,4 +213,4 @@ def _calculate_expectation_values(
         * (check_parity_of_vector(bitstrings, [op[0] for op in term]) * 2 - 1)
         for term, coefficient in operator.terms.items()
     ]
-    return np.array(expectation_values_list).sum(axis=0)
+    return np.array(np.array(expectation_values_list).sum(axis=0))
```
[the remark this answers is not in the document]
Comment (mstechly) [reply src/python/zquantum/qaoa/estimators/_cvar_estimator.py:216]: Well, the problem is that `np.sum` returns "`np.array` or `number`", but our function should return just `np.array`. 
So I added this one `np.array` more to get rid of mypy issue.

**Review by mstechly — APPROVED** [review]

**Review by dexter2206 — APPROVED** [review]

**Pull request merged**